### PR TITLE
fix: reject transactions with insufficient funds

### LINKED ISSUE #1831: Ganache should fail with `insufficient funds for transfer` error when a transaction is added to the pool with insufficient funds

### PR DESCRIPTION
**Pull request by davidmurdoch**
fixes #1831

### CODE REVIEW

File: src/chains/ethereum/ethereum/src/transaction-pool.ts [hunk 1]
```diff
@@ -204,6 +207,22 @@ export default class TransactionPool extends Emittery.Typed<{}, "drain"> {
...
+        this.#accountPromises.delete(origin);
+      });
+    }
+    const transactor = await transactorPromise;
+
+    const cost =
+      transaction.gas.toBigInt() * transaction.maxGasPrice().toBigInt() +
+      transaction.value.toBigInt();
+    if (transactor.balance.toBigInt() < cost) {
+      throw new Error("insufficient funds for gas * price + value");
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/src/transaction-pool.ts]: Can this be a coded error?

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts [hunk 2]
```diff
@@ -47,24 +47,58 @@ describe("api", () => {
       });
 
       describe("insufficient funds", () => {
-        it("returns an error when account has insufficient funds to send the transaction", async () => {
-          const p = await getProvider({
+        const types = ["0x0", "0x1", "0x2"] as const;
+        it("returns an error when account has insufficient funds to transfer the value", async () => {
+          const gasCost = 99967968750001;
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]: Does this really work? I'd be surprised if all three tx types have the same gas cost.

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts [hunk 3]
```diff
@@ -47,24 +47,58 @@ describe("api", () => {
...
+              const balance = parseInt(await getBalance(from), 16);
+              // fire a transaction without awaiting it in order to spend some
+              // gas
+              const tx = { type: types[i], from, to };
+              sendTx(tx);
+              await assert.rejects(
+                sendTx({
+                  ...tx,
+                  // attempt to zero out the account. this tx will fail because
+                  // the previous (pending transaction) will spend some it's
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]:
```suggestion
                  // the previous (pending transaction) will spend some of its
```

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts [hunk 4]
```diff
@@ -47,24 +47,58 @@ describe("api", () => {
       });
 
       describe("insufficient funds", () => {
-        it("returns an error when account has insufficient funds to send the transaction", async () => {
-          const p = await getProvider({
+        const types = ["0x0", "0x1", "0x2"] as const;
+        it("returns an error when account has insufficient funds to transfer the value", async () => {
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]: I feel like we should update the "it" text for this and the next test to differentiate the two - one is rejected by the tx pool, the other by the VM

File: src/chains/ethereum/ethereum/tests/transaction-pool.test.ts [hunk 5]
```diff
@@ -81,8 +81,12 @@ describe("transaction pool", async () => {
     // returns an account's nonce
     blockchain = {
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/tests/transaction-pool.test.ts:83]: It kinda seems like we should add a test for the insufficient balance rejection here, right? To get all code paths

**Review by MicaiahReid — CHANGES_REQUESTED** [review]

File: src/chains/ethereum/ethereum/tests/transaction-pool.test.ts (hunk 5 above)
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/tests/transaction-pool.test.ts:83]: added!

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts (hunk 4 above)
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]: changed!

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts (hunk 3 above)
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]: fixed

File: src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts (hunk 2 above)
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/tests/api/eth/sendTransaction.test.ts]: it just has to be close enough. i've updated the var name to `approximateGasCost`

File: src/chains/ethereum/ethereum/src/transaction-pool.ts (hunk 1 above)
Comment (davidmurdoch) [reply src/chains/ethereum/ethereum/src/transaction-pool.ts]: done!

**Review by MicaiahReid — APPROVED** [review]

**Pull request merged**